**Scripts and reference genomes used in "Mechanism of homology search expansion during recombinational DNA repair", Dumont et al. Molecular Cell 2024**

This repository provides annotated shell scripts, Jupyter notebooks, reference genomes, and input files to produce the Hi-C and ssHi-C data from paired-end fastq files. It requires Python 3.9 and the following softwares:
ssHicstuff (available at https://github.com/nmendiboure/sshicstuff)
Hicstuff (version 3.1.2 available at https://github.com/baudrly/hicstuff)
Chromosight (version 1.6.3, available at https://github.com/koszullab/chromosight)
Serpentine (version 0.1.3, available at https://github.com/koszullab/serpentine)
Tinycov (version 0.3.1, available at https://github.com/cmdoret/tinycov)
TinyMapper (version 0.1, available at https://github.com/js2264/tinyMapper)
Bowtie2 (version 2.3.5.1  available online at http://bowtie-bio.sourceforge.net/bowtie2/) 
Samtools (version 1.3.1 available online at https://github.com/samtools/samtools) 
Bedtools (version 2.27.1 available online at https://github.com/arq5x/bedtools2) 
Cooler (version 0.9.1 available online at https://github.com/open2c/cooler)
