{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import sys\n",
    "import pandas as pd\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the data: fragment files (bin 1kb); chromosomes coordinates; matrice in dense format (bin 1kb). The data are loaded with pandas, and the matrice is converted to an array with numpy so as to access specific data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_seqs= \"/mnt/d/Science/OwnCloud/Sequences/Genomes/\"\n",
    "out= \"/mnt/d/Science/OwnCloud/Experiments/Capture/Quantif/\"\n",
    "input_dir = \"/mnt/e/Science/ssHiC/Contact_files/\"\n",
    "\n",
    "#specify the path to the tab-delimited file containing chromosome name, length, and centromere location\n",
    "\n",
    "#chrs = pd.read_csv(str(dir_seqs + \"Yeast_genome_W303_Marcand/W303_710_chromosome_centromere_coordinates.tsv\"), sep = \"\\t\")\n",
    "#chrs = pd.read_csv(str(dir_seqs + \"Yeast_Genome_SK1_2019/sk1_chromosomes_centro_chr4_v1.tsv\"), sep = \"\\t\")\n",
    "chrs = pd.read_csv(str(dir_seqs + \"Yeast_genome_S288c_R64-2-1_2014/S288c_chr_centro_coordinates.tsv\"), sep = \"\\t\")\n",
    "#print(chrs)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the data. It will show an array of integer (if the matrice is not pre-normalized). Sometimes tsv or csv. check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      Chr   Length  Left_arm_length  Right_arm_length\n",
      "0    chr1   230218           151465             78753\n",
      "1    chr2   813184           238207            574977\n",
      "2    chr3   316620           114385            202235\n",
      "3    chr4  1531933           449711           1082222\n",
      "4    chr5   576874           151987            424887\n",
      "5    chr6   270161           148510            121651\n",
      "6    chr7  1090940           497038            593902\n",
      "7    chr8   562643           105703            456940\n",
      "8    chr9   439888           355629             84259\n",
      "9   chr10   745751           436425            309326\n",
      "10  chr11   666816           440246            226570\n",
      "11  chr12  1078177           150947            927230\n",
      "12  chr13   924431           268031            656400\n",
      "13  chr14   784333           628758            155575\n",
      "14  chr15  1091291           326702            764589\n",
      "15  chr16   948066           555957            392109\n"
     ]
    }
   ],
   "source": [
    "print(chrs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Normalize the data in the matrice by the total number of reads. It will return an array of floats."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[16 55 31 ...  0  0  0]\n",
      " [55  5 57 ...  0  0  0]\n",
      " [31 57 15 ...  0  0  0]\n",
      " ...\n",
      " [ 0  0  0 ...  6  4  7]\n",
      " [ 0  0  0 ...  4  0  0]\n",
      " [ 0  0  0 ...  7  0  0]]\n",
      "total number of contacts in the matrix = 9300702\n",
      "--------------------------\n",
      "done\n",
      "--------------------------\n",
      "[0.7058082531802428, 0.8031741773834581, 0.7441081671201086, 0.8293291044141401, 0.7581818392751373, 0.7117608791732809, 0.7936748287783904, 0.7705047598194028, 0.7412714405875377, 0.7806330460841713, 0.7757567401115862, 0.8222550741525324, 0.7919848609276061, 0.7833170402599534, 0.8009147913991231, 0.7832200524000144]\n",
      "done\n",
      "--------------------------\n",
      "done\n",
      "--------------------------\n",
      "[2.1288715625981783e-05, 2.4514278599615386e-05, 2.236390328385965e-05, 2.075112179704285e-05, 2.010600920231613e-05, 2.5374428725917677e-05, 2.2471422049647433e-05, 2.5912022554856612e-05, 3.773908679151316e-05, 2.741728527588563e-05, 2.365412847331309e-05, 2.107367809440621e-05, 2.096615932861842e-05, 2.1181196860193995e-05, 2.4406759833827598e-05]\n",
      "done\n",
      "--------------------------\n",
      "done\n",
      "--------------------------\n",
      "done\n",
      "[[14 44 24 ...  0  0  0]\n",
      " [44  8 79 ...  0  0  0]\n",
      " [24 79  6 ...  0  0  0]\n",
      " ...\n",
      " [ 0  0  0 ...  1  4  4]\n",
      " [ 0  0  0 ...  4  3  3]\n",
      " [ 0  0  0 ...  4  3  0]]\n",
      "total number of contacts in the matrix = 7818460\n",
      "--------------------------\n",
      "done\n",
      "--------------------------\n",
      "[0.7882988953078753, 0.8795614916813289, 0.8115692738942687, 0.8985258793093396, 0.8379656735189993, 0.7920876602114327, 0.8683192808338136, 0.8481662921900525, 0.8222595882144297, 0.8524030971611812, 0.8514995911692775, 0.9002848443475243, 0.8660394693699216, 0.8569581628920228, 0.8743411550718792, 0.8611847343188341]\n",
      "done\n",
      "--------------------------\n",
      "done\n",
      "--------------------------\n",
      "[1.8290046889029297e-05, 2.238292451454634e-05, 1.777843718583966e-05, 1.5092486244094107e-05, 1.765053476004226e-05, 2.2127119662951524e-05, 2.199921723715411e-05, 2.097599783077485e-05, 3.00570700623908e-05, 2.0336485701787817e-05, 1.7778437185839662e-05, 1.84179493148267e-05, 2.2510826940343742e-05, 2.1231802682369672e-05, 2.0720192979180035e-05]\n",
      "done\n",
      "--------------------------\n",
      "done\n",
      "--------------------------\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "samples = [\"AD301_S288c_DSB_LY_Capture_artificial_cutsite_q30\",\n",
    "          \"AD302_S288c_DSB_LY_Capture_artificial_cutsite_q30\"]\n",
    "\n",
    "for sample in samples:\n",
    "    frags = pd.read_csv(str(input_dir + \"Rebin/S288c/\" + sample + \"_1kb.frag.tsv\"), sep = \"\\t\")\n",
    "    mat = pd.read_csv(str(input_dir + \"Dense/S288c/\" + sample + \"_dense_1kb.mat.csv\"), sep = \" \", header = None)\n",
    "    mat = np.array(mat)\n",
    "    print(mat)\n",
    "    #frags = frags.replace(\"chr4_v1\", \"chr4\") \n",
    "\n",
    "    #print(sum(mat[:,:]))\n",
    "    print(\"total number of contacts in the matrix =\", sum(sum(mat)))\n",
    "    mat_norm=mat/sum(sum(mat)) #matrix normalized\n",
    "\n",
    "\n",
    "    #Define the lines in frags corresponding to each chromosomes in chrs.\n",
    "    #Define the lines in frags of start and end positions of fragments corresponding to centromeres [-X kb to +X kb] around the cen-containing bin (= 5 bins total)\n",
    "    #Define the lines in frags of the start and end positions of each telomere [consider a whole 20kb region]\n",
    "    #chromosomes\n",
    "    coord_chr_start = []\n",
    "    coord_chr_end = []\n",
    "    i = 0\n",
    "    for i in chrs.loc[: ,\"Chr\"]:\n",
    "        coord_chr_start.append(min(frags.loc[frags[\"chrom\"]== str(i)].index.values))\n",
    "        coord_chr_end.append(max(frags.loc[frags[\"chrom\"]== str(i)].index.values))\n",
    "    #print(coord_chr_start)\n",
    "    #print(coord_chr_end)\n",
    "\n",
    "    #centromeres\n",
    "    #extra_bin = input(\"extra kb on each side of the centromeric bin?\") #Normally 2 to have a 5kb window\n",
    "    extra_bin = 14\n",
    "    coord_cen_start = []\n",
    "    coord_cen_end = []\n",
    "    i = 0\n",
    "    for i in chrs.loc[: ,\"Chr\"]:\n",
    "        cen = int(chrs.loc[chrs[\"Chr\"] == str(i)][\"Left_arm_length\"])\n",
    "        cen_row=int(frags.loc[(frags[\"chrom\"]== str(i)) & (frags[\"start_pos\"] < cen) & (frags[\"end_pos\"] > cen)].index.values)\n",
    "        coord_cen_start.append(cen_row - int(extra_bin))\n",
    "        coord_cen_end.append(cen_row + int(extra_bin) + 1)\n",
    "    #print(coord_cen_start)\n",
    "    #print(coord_cen_end)\n",
    "\n",
    "\n",
    "    print(\"--------------------------\")\n",
    "    #Quantif telomeres contacts\n",
    "    coord_tel_start = []\n",
    "    coord_tel_end = []\n",
    "    i = 0\n",
    "    for i in chrs.loc[: ,\"Chr\"]:\n",
    "        tel1 = 1+2\n",
    "        tel2 = int(chrs.loc[chrs[\"Chr\"] == str(i)][\"Length\"]) -1000\n",
    "        tel1_row=int(frags.loc[(frags[\"chrom\"]== str(i)) & (frags[\"start_pos\"] < tel1) & (frags[\"end_pos\"] > tel1)].index.values)\n",
    "        coord_tel_start.append(tel1_row)\n",
    "        coord_tel_end.append(tel1_row + 19)\n",
    "        tel2_row=int(frags.loc[(frags[\"chrom\"]== str(i)) & (frags[\"start_pos\"] < tel2) & (frags[\"end_pos\"] > tel2)].index.values)\n",
    "        coord_tel_start.append(tel2_row - 19)\n",
    "        coord_tel_end.append(tel2_row)\n",
    "    #print(coord_tel_start)\n",
    "    #print(coord_tel_end)\n",
    "    print(\"done\")\n",
    "\n",
    "\n",
    "    #QUANTIF INTRA CONTACTS FOR EACH CHROMOSOMES\n",
    "    print(\"--------------------------\")\n",
    "    output_dir = out + \"Intra/\"\n",
    "    intra_contacts = []\n",
    "    i=0\n",
    "    while i < len(coord_chr_start):\n",
    "        start = coord_chr_start[i]\n",
    "        end = coord_chr_end[i]\n",
    "        intra=0\n",
    "        intra=sum(sum(mat_norm[start : end , start : end]))\n",
    "        all_contacts=sum(sum(mat_norm[start : end , ]))\n",
    "        intra_contacts.append(intra/all_contacts)\n",
    "        i += 1\n",
    "    print(intra_contacts)\n",
    "    print(\"done\")\n",
    "    np.savetxt(output_dir + str(sample) + \"_intra.tsv\", intra_contacts, delimiter = \"\\t\")\n",
    "\n",
    "\n",
    "\n",
    "    print(\"--------------------------\")\n",
    "    #Quantify the contacts between the centromeres\n",
    "    output_dir = out + \"Centro/\"\n",
    "    cen_contacts = []\n",
    "    i=0\n",
    "    while i < len(coord_cen_start):\n",
    "        j=0\n",
    "        while j < len(coord_cen_start):\n",
    "            if i != j:\n",
    "                start1 = coord_cen_start[i]\n",
    "                end1 = coord_cen_end[i]\n",
    "                start2 = coord_cen_start[j]\n",
    "                end2 = coord_cen_end[j]\n",
    "                cen_contacts.append(sum(sum(mat_norm[start1 : end1 , start2 : end2])))\n",
    "            j += 1\n",
    "        i += 1\n",
    "    #print(cen_contacts)\n",
    "    print(\"done\")\n",
    "    np.savetxt(output_dir + str(sample) + \"_centro_binsize_\" + str(int(extra_bin) * 2 + 1) + \"kb.tsv\", cen_contacts, delimiter = \"\\t\")\n",
    "\n",
    "\n",
    "\n",
    "    print(\"--------------------------\")\n",
    "    #Quantify the contacts between CEN5 and other 15 centromeres\n",
    "    output_dir = out + \"Centro/\"\n",
    "    cen_contacts_chr5 = []\n",
    "    i = 5\n",
    "    j = 0\n",
    "    while j < len(coord_cen_start):\n",
    "        if i != j:\n",
    "            start1 = coord_cen_start[i]\n",
    "            end1 = coord_cen_end[i]\n",
    "            start2 = coord_cen_start[j]\n",
    "            end2 = coord_cen_end[j]\n",
    "            cen_contacts_chr5.append(sum(sum(mat_norm[start1 : end1 , start2 : end2])))\n",
    "        j += 1\n",
    "\n",
    "    print(cen_contacts_chr5)\n",
    "    print(\"done\")\n",
    "    np.savetxt(output_dir + str(sample) + \"_chr5_centro_binsize_\" + str(int(extra_bin) * 2 + 1) + \"kb.tsv\", cen_contacts_chr5, delimiter = \"\\t\")\n",
    "\n",
    "\n",
    "\n",
    "    print(\"--------------------------\")\n",
    "    #Quantify contcts between telomeres\n",
    "    output_dir = out + \"Telo/\"\n",
    "    tel_contacts = []\n",
    "    i=0\n",
    "    while i < len(coord_tel_start):\n",
    "        j=0\n",
    "        while j < len(coord_tel_end):\n",
    "            if i != j:\n",
    "                a = 0\n",
    "                start1 = coord_tel_start[i]\n",
    "                end1 = coord_tel_end[i]\n",
    "                start2 = coord_tel_start[j]\n",
    "                end2 = coord_tel_end[j]\n",
    "                a = sum(sum(mat_norm[start1 : end1 , start2 : end2]))\n",
    "                if a != 0:\n",
    "                    tel_contacts.append(a)\n",
    "            j += 1\n",
    "        i += 1\n",
    "    #print(tel_contacts)\n",
    "    print(\"done\")\n",
    "    np.savetxt(output_dir + str(sample) + \"_telo_20kb.tsv\", tel_contacts, delimiter = \"\\t\")\n",
    "\n",
    "\n",
    "    print(\"--------------------------\")\n",
    "    #Quantify contacts between telomeres of chr5 and other telomeres\n",
    "    output_dir = out + \"Telo/\"\n",
    "    tel_contacts_chr5 = []\n",
    "    for i in [11,12]:\n",
    "        j=0\n",
    "        while j < len(coord_tel_end):\n",
    "            if i != j:\n",
    "                start1 = coord_tel_start[i]\n",
    "                end1 = coord_tel_end[i]\n",
    "                start2 = coord_tel_start[j]\n",
    "                end2 = coord_tel_end[j]\n",
    "                tel_contacts_chr5.append(sum(sum(mat_norm[start1 : end1 , start2 : end2])))\n",
    "            j += 1\n",
    "    #print(tel_contacts_chr5)\n",
    "    print(\"done\")\n",
    "    np.savetxt(output_dir + str(sample) + \"_chr5_telo_10kb.tsv\", tel_contacts_chr5, delimiter = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr4_v1\n"
     ]
    }
   ],
   "source": [
    "print(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For DpnII only matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " ...\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]]\n",
      "total number of contacts in the matrix = 44095939\n",
      "--------------------------\n",
      "done\n",
      "--------------------------\n",
      "[0.5718245878571194, 0.6095497736679477, 0.5841790820689022, 0.6707477657610017, 0.5805378969287079, 0.5804405050205328, 0.63006794811416, 0.5948712942568966, 0.585256185075522, 0.6198645545646377, 0.6052738830233043, 0.6756934306770003, 0.6268063410157241, 0.6071886689501065, 0.6353966211783003, 0.6161534180572109]\n",
      "done\n",
      "--------------------------\n",
      "done\n",
      "--------------------------\n",
      "[3.610309783855606e-05, 3.440226094289544e-05, 3.007079631527974e-05, 3.270142404723483e-05, 3.524134047808801e-05, 3.1182009753778006e-05, 3.583096393525037e-05, 3.7894646035318584e-05, 3.596703088690323e-05, 3.8892470347439476e-05, 3.1499499307634665e-05, 3.24519679692046e-05, 3.4084771389038794e-05, 3.594435306162773e-05, 3.0501674995513756e-05]\n",
      "done\n",
      "--------------------------\n",
      "done\n",
      "--------------------------\n",
      "done\n",
      "[[0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " ...\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]\n",
      " [0 0 0 ... 0 0 0]]\n",
      "total number of contacts in the matrix = 58931547\n",
      "--------------------------\n",
      "done\n",
      "--------------------------\n",
      "[0.768714186338129, 0.7925346442861313, 0.7732902406405997, 0.8221137625283472, 0.7805739461758876, 0.7759758549532277, 0.8032571963034428, 0.7855949847587248, 0.7751464863946875, 0.7985786821556321, 0.8000011358005769, 0.8289985662006332, 0.8033202213066368, 0.7927398584031411, 0.8091625802359796, 0.7940956550703728]\n",
      "done\n",
      "--------------------------\n",
      "done\n",
      "--------------------------\n",
      "[6.312408530527801e-06, 4.496742635994266e-06, 5.837281006724631e-06, 4.5137114761300935e-06, 5.7863744863171505e-06, 4.632493357080885e-06, 6.838442574738451e-06, 7.279632418269963e-06, 5.243371601970673e-06, 6.19362664957701e-06, 5.463966523736431e-06, 4.293116554364339e-06, 5.9221252074037705e-06, 5.039745520340743e-06, 4.360991914907647e-06]\n",
      "done\n",
      "--------------------------\n",
      "done\n",
      "--------------------------\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "#samples = [\"AD106\", \"AD107\", \"AD108\", \"AD109\", \"AD110\", \"AD111\", \"AD112\", \"AD113\"]\n",
    "#samples = [\"AP51\", \"AP52\"]\n",
    "samples = [\"AP36_AP37_merged\", \"AP38_AP39_merged\"]\n",
    "\n",
    "for sample in samples:\n",
    "    frags = pd.read_csv(str(input_dir + \"Rebin/\" + sample + \"_DpnII_1kb.frag.tsv\"), sep = \"\\t\")\n",
    "    mat = pd.read_csv(str(input_dir + \"Dense/\" + sample + \"_dense_DpnII_1kb.mat.csv\"), sep = \" \", header = None)\n",
    "    mat = np.array(mat)\n",
    "    print(mat)\n",
    "    #frags = frags.replace(\"chr4_v1\", \"chr4\") \n",
    "\n",
    "    #print(sum(mat[:,:]))\n",
    "    print(\"total number of contacts in the matrix =\", sum(sum(mat)))\n",
    "    mat_norm=mat/sum(sum(mat)) #matrix normalized\n",
    "\n",
    "\n",
    "    #Define the lines in frags corresponding to each chromosomes in chrs.\n",
    "    #Define the lines in frags of start and end positions of fragments corresponding to centromeres [-X kb to +X kb] around the cen-containing bin (= 5 bins total)\n",
    "    #Define the lines in frags of the start and end positions of each telomere [consider a whole 20kb region]\n",
    "    #chromosomes\n",
    "    coord_chr_start = []\n",
    "    coord_chr_end = []\n",
    "    i = 0\n",
    "    for i in chrs.loc[: ,\"Chr\"]:\n",
    "        coord_chr_start.append(min(frags.loc[frags[\"chrom\"]== str(i)].index.values))\n",
    "        coord_chr_end.append(max(frags.loc[frags[\"chrom\"]== str(i)].index.values))\n",
    "    #print(coord_chr_start)\n",
    "    #print(coord_chr_end)\n",
    "\n",
    "    #centromeres\n",
    "    #extra_bin = input(\"extra kb on each side of the centromeric bin?\") #Normally 2 to have a 5kb window\n",
    "    extra_bin = 14\n",
    "    coord_cen_start = []\n",
    "    coord_cen_end = []\n",
    "    i = 0\n",
    "    for i in chrs.loc[: ,\"Chr\"]:\n",
    "        cen = int(chrs.loc[chrs[\"Chr\"] == str(i)][\"Left_arm_length\"])\n",
    "        cen_row=int(frags.loc[(frags[\"chrom\"]== str(i)) & (frags[\"start_pos\"] < cen) & (frags[\"end_pos\"] > cen)].index.values)\n",
    "        coord_cen_start.append(cen_row - int(extra_bin))\n",
    "        coord_cen_end.append(cen_row + int(extra_bin) + 1)\n",
    "    #print(coord_cen_start)\n",
    "    #print(coord_cen_end)\n",
    "\n",
    "\n",
    "    print(\"--------------------------\")\n",
    "    #Quantif telomeres contacts\n",
    "    coord_tel_start = []\n",
    "    coord_tel_end = []\n",
    "    i = 0\n",
    "    for i in chrs.loc[: ,\"Chr\"]:\n",
    "        tel1 = 1+2\n",
    "        tel2 = int(chrs.loc[chrs[\"Chr\"] == str(i)][\"Length\"]) -1000\n",
    "        tel1_row=int(frags.loc[(frags[\"chrom\"]== str(i)) & (frags[\"start_pos\"] < tel1) & (frags[\"end_pos\"] > tel1)].index.values)\n",
    "        coord_tel_start.append(tel1_row)\n",
    "        coord_tel_end.append(tel1_row + 19)\n",
    "        tel2_row=int(frags.loc[(frags[\"chrom\"]== str(i)) & (frags[\"start_pos\"] < tel2) & (frags[\"end_pos\"] > tel2)].index.values)\n",
    "        coord_tel_start.append(tel2_row - 19)\n",
    "        coord_tel_end.append(tel2_row)\n",
    "    #print(coord_tel_start)\n",
    "    #print(coord_tel_end)\n",
    "    print(\"done\")\n",
    "\n",
    "\n",
    "    #QUANTIF INTRA CONTACTS FOR EACH CHROMOSOMES\n",
    "    print(\"--------------------------\")\n",
    "    output_dir = out + \"Intra/\"\n",
    "    intra_contacts = []\n",
    "    i=0\n",
    "    while i < len(coord_chr_start):\n",
    "        start = coord_chr_start[i]\n",
    "        end = coord_chr_end[i]\n",
    "        intra=0\n",
    "        intra=sum(sum(mat_norm[start : end , start : end]))\n",
    "        all_contacts=sum(sum(mat_norm[start : end , ]))\n",
    "        intra_contacts.append(intra/all_contacts)\n",
    "        i += 1\n",
    "    print(intra_contacts)\n",
    "    print(\"done\")\n",
    "    np.savetxt(output_dir + str(sample) + \"_intra.tsv\", intra_contacts, delimiter = \"\\t\")\n",
    "\n",
    "\n",
    "\n",
    "    print(\"--------------------------\")\n",
    "    #Quantify the contacts between the centromeres\n",
    "    output_dir = out + \"Centro/\"\n",
    "    cen_contacts = []\n",
    "    i=0\n",
    "    while i < len(coord_cen_start):\n",
    "        j=0\n",
    "        while j < len(coord_cen_start):\n",
    "            if i != j:\n",
    "                start1 = coord_cen_start[i]\n",
    "                end1 = coord_cen_end[i]\n",
    "                start2 = coord_cen_start[j]\n",
    "                end2 = coord_cen_end[j]\n",
    "                cen_contacts.append(sum(sum(mat_norm[start1 : end1 , start2 : end2])))\n",
    "            j += 1\n",
    "        i += 1\n",
    "    #print(cen_contacts)\n",
    "    print(\"done\")\n",
    "    np.savetxt(output_dir + str(sample) + \"_centro_binsize_\" + str(int(extra_bin) * 2 + 1) + \"kb.tsv\", cen_contacts, delimiter = \"\\t\")\n",
    "\n",
    "\n",
    "\n",
    "    print(\"--------------------------\")\n",
    "    #Quantify the contacts between CEN5 and other 15 centromeres\n",
    "    output_dir = out + \"Centro/\"\n",
    "    cen_contacts_chr5 = []\n",
    "    i = 5\n",
    "    j = 0\n",
    "    while j < len(coord_cen_start):\n",
    "        if i != j:\n",
    "            start1 = coord_cen_start[i]\n",
    "            end1 = coord_cen_end[i]\n",
    "            start2 = coord_cen_start[j]\n",
    "            end2 = coord_cen_end[j]\n",
    "            cen_contacts_chr5.append(sum(sum(mat_norm[start1 : end1 , start2 : end2])))\n",
    "        j += 1\n",
    "\n",
    "    print(cen_contacts_chr5)\n",
    "    print(\"done\")\n",
    "    np.savetxt(output_dir + str(sample) + \"_chr5_centro_binsize_\" + str(int(extra_bin) * 2 + 1) + \"kb.tsv\", cen_contacts_chr5, delimiter = \"\\t\")\n",
    "\n",
    "\n",
    "\n",
    "    print(\"--------------------------\")\n",
    "    #Quantify contcts between telomeres\n",
    "    output_dir = out + \"Telo/\"\n",
    "    tel_contacts = []\n",
    "    i=0\n",
    "    while i < len(coord_tel_start):\n",
    "        j=0\n",
    "        while j < len(coord_tel_end):\n",
    "            if i != j:\n",
    "                a = 0\n",
    "                start1 = coord_tel_start[i]\n",
    "                end1 = coord_tel_end[i]\n",
    "                start2 = coord_tel_start[j]\n",
    "                end2 = coord_tel_end[j]\n",
    "                a = sum(sum(mat_norm[start1 : end1 , start2 : end2]))\n",
    "                if a != 0:\n",
    "                    tel_contacts.append(a)\n",
    "            j += 1\n",
    "        i += 1\n",
    "    #print(tel_contacts)\n",
    "    print(\"done\")\n",
    "    np.savetxt(output_dir + str(sample) + \"_telo_20kb.tsv\", tel_contacts, delimiter = \"\\t\")\n",
    "\n",
    "\n",
    "    print(\"--------------------------\")\n",
    "    #Quantify contacts between telomeres of chr5 and other telomeres\n",
    "    output_dir = out + \"Telo/\"\n",
    "    tel_contacts_chr5 = []\n",
    "    for i in [11,12]:\n",
    "        j=0\n",
    "        while j < len(coord_tel_end):\n",
    "            if i != j:\n",
    "                start1 = coord_tel_start[i]\n",
    "                end1 = coord_tel_end[i]\n",
    "                start2 = coord_tel_start[j]\n",
    "                end2 = coord_tel_end[j]\n",
    "                tel_contacts_chr5.append(sum(sum(mat_norm[start1 : end1 , start2 : end2])))\n",
    "            j += 1\n",
    "    #print(tel_contacts_chr5)\n",
    "    print(\"done\")\n",
    "    np.savetxt(output_dir + str(sample) + \"_chr5_telo_10kb.tsv\", tel_contacts_chr5, delimiter = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "828469\n",
      "chr2\n",
      "       id chrom  start_pos  end_pos   size  gc_content\n",
      "229     1  chr2          0     1000   1000    0.460091\n",
      "230     2  chr2       1000     2000   1000    0.438095\n",
      "231     3  chr2       2000     3000   1000    0.451932\n",
      "232     4  chr2       3000     4000   1000    0.452892\n",
      "233     5  chr2       4000     5000   1000    0.389157\n",
      "...   ...   ...        ...      ...    ...         ...\n",
      "1053  825  chr2     824000   804833 -19167    0.472953\n",
      "1054  826  chr2     825000   804833 -20167    0.506568\n",
      "1055  827  chr2     826000   805833 -20167         NaN\n",
      "1056  828  chr2     827000   806833 -20167         NaN\n",
      "1057  829  chr2     828000   804833 -23167    0.467177\n",
      "\n",
      "[829 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "print(tel1)\n",
    "print(tel2)\n",
    "print(str(i))\n",
    "print(frags.loc[(frags[\"chrom\"]== str(i)) & (frags[\"start_pos\"] < tel2)])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define the lines in frags corresponding to each chromosomes in chrs.\n",
    "\n",
    "Define the lines in frags of start and end positions of fragments corresponding to centromeres [-X kb to +X kb] around the cen-containing bin (= 5 bins total)\n",
    "\n",
    "Define the lines in frags of the start and end positions of each telomere [consider a whole 20kb region]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "#chromosomes\n",
    "coord_chr_start = []\n",
    "coord_chr_end = []\n",
    "i = 0\n",
    "for i in range(0, len(chrs)):\n",
    "    coord_chr_start.append(min(frags.loc[frags[\"chrom\"]== \"chr\" + str(i + 1)].index.values))\n",
    "    coord_chr_end.append(max(frags.loc[frags[\"chrom\"]== \"chr\" + str(i + 1)].index.values))\n",
    "#print(coord_chr_start)\n",
    "#print(coord_chr_end)\n",
    "\n",
    "#centromeres\n",
    "#extra_bin = input(\"extra kb on each side of the centromeric bin?\") #Normally 2 to have a 5kb window\n",
    "extra_bin = 2\n",
    "coord_cen_start = []\n",
    "coord_cen_end = []\n",
    "i = 0\n",
    "for i in range(0, len(chrs)):\n",
    "    cen = int(chrs.loc[chrs[\"Chr\"] == \"chr\" + str(i+1)][\"Left_arm_length\"])\n",
    "    cen_row=int(frags.loc[(frags[\"chrom\"]== \"chr\" + str(i + 1)) & (frags[\"start_pos\"] < cen) & (frags[\"end_pos\"] > cen)].index.values)\n",
    "    coord_cen_start.append(cen_row - int(extra_bin))\n",
    "    coord_cen_end.append(cen_row + int(extra_bin) + 1)\n",
    "#print(coord_cen_start)\n",
    "#print(coord_cen_end)\n",
    "\n",
    "#telomeres\n",
    "coord_tel_start = []\n",
    "coord_tel_end = []\n",
    "i = 0\n",
    "for i in range(0, len(chrs)):\n",
    "    tel1 = 1+2\n",
    "    tel2 = int(chrs.loc[chrs[\"Chr\"] == \"chr\" + str(i+1)][\"Length\"]) -1000\n",
    "    tel1_row=int(frags.loc[(frags[\"chrom\"]== \"chr\" + str(i + 1)) & (frags[\"start_pos\"] < tel1) & (frags[\"end_pos\"] > tel1)].index.values)\n",
    "    coord_tel_start.append(tel1_row)\n",
    "    coord_tel_end.append(tel1_row + 19)\n",
    "    tel2_row=int(frags.loc[(frags[\"chrom\"]== \"chr\" + str(i + 1)) & (frags[\"start_pos\"] < tel2) & (frags[\"end_pos\"] > tel2)].index.values)\n",
    "    coord_tel_start.append(tel2_row - 19)\n",
    "    coord_tel_end.append(tel2_row)\n",
    "#print(coord_tel_start)\n",
    "#print(coord_tel_end)\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quantify the intra-chromosomal contacts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.7114501551015645, 0.8076396302773152, 0.7419116190507897, 0.8468769415228061, 0.7728027290857843, 0.7270359803152728, 0.81747645369735, 0.7746843288935392, 0.756927915447601, 0.7938471992285933, 0.7906909439703667, 0.8323624270277716, 0.8113514389111354, 0.8008498666969448, 0.816766416005933, 0.8057428076145948]\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "output_dir = out + \"Intra/\"\n",
    "intra_contacts = []\n",
    "i=0\n",
    "while i < len(coord_chr_start):\n",
    "    start = coord_chr_start[i]\n",
    "    end = coord_chr_end[i]\n",
    "    intra=0\n",
    "    intra=sum(sum(mat_norm[start : end , start : end]))\n",
    "    all_contacts=sum(sum(mat_norm[start : end , ]))\n",
    "    intra_contacts.append(intra/all_contacts)\n",
    "    i += 1\n",
    "print(intra_contacts)\n",
    "print(\"done\")\n",
    "np.savetxt(output_dir + str(sample) + \"_intra.tsv\", intra_contacts, delimiter = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quantify the contacts between the centromeres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "output_dir = out + \"Centro/\"\n",
    "cen_contacts = []\n",
    "i=0\n",
    "while i < len(coord_cen_start):\n",
    "    j=0\n",
    "    while j < len(coord_cen_start):\n",
    "        if i != j:\n",
    "            start1 = coord_cen_start[i]\n",
    "            end1 = coord_cen_end[i]\n",
    "            start2 = coord_cen_start[j]\n",
    "            end2 = coord_cen_end[j]\n",
    "            cen_contacts.append(sum(sum(mat_norm[start1 : end1 , start2 : end2])))\n",
    "        j += 1\n",
    "    i += 1\n",
    "#print(cen_contacts)\n",
    "print(\"done\")\n",
    "np.savetxt(output_dir + str(sample) + \"_centro_binsize_\" + str(int(extra_bin) * 2 + 1) + \"kb.tsv\", cen_contacts, delimiter = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quantify the contact of Cen 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1.98174805151351e-06, 2.7891268873153096e-06, 2.1285442034774737e-06, 2.079612152822819e-06, 2.006214076840837e-06, 2.177476254132128e-06, 2.4466025327327283e-06, 2.2753403554414376e-06, 1.8838839502042009e-06, 2.1040781781501465e-06, 1.467961519639637e-06, 1.467961519639637e-06, 2.0796121528228192e-06, 2.1040781781501465e-06, 2.0796121528228192e-06]\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "output_dir = out + \"Centro/\"\n",
    "cen_contacts_chr5 = []\n",
    "i = 5\n",
    "j = 0\n",
    "while j < len(coord_cen_start):\n",
    "    if i != j:\n",
    "        start1 = coord_cen_start[i]\n",
    "        end1 = coord_cen_end[i]\n",
    "        start2 = coord_cen_start[j]\n",
    "        end2 = coord_cen_end[j]\n",
    "        cen_contacts_chr5.append(sum(sum(mat_norm[start1 : end1 , start2 : end2])))\n",
    "    j += 1\n",
    "\n",
    "print(cen_contacts_chr5)\n",
    "print(\"done\")\n",
    "np.savetxt(output_dir + str(sample) + \"_chr5_centro_binsize_\" + str(int(extra_bin) * 2 + 1) + \"kb.tsv\", cen_contacts_chr5, delimiter = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quantify the contacts between telomeres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "output_dir = out + \"Telo/\"\n",
    "tel_contacts = []\n",
    "i=0\n",
    "while i < len(coord_tel_start):\n",
    "    j=0\n",
    "    while j < len(coord_tel_end):\n",
    "        if i != j:\n",
    "            a = 0\n",
    "            start1 = coord_tel_start[i]\n",
    "            end1 = coord_tel_end[i]\n",
    "            start2 = coord_tel_start[j]\n",
    "            end2 = coord_tel_end[j]\n",
    "            a = sum(sum(mat_norm[start1 : end1 , start2 : end2]))\n",
    "            if a != 0:\n",
    "                tel_contacts.append(a)\n",
    "        j += 1\n",
    "    i += 1\n",
    "#print(tel_contacts)\n",
    "print(\"done\")\n",
    "np.savetxt(output_dir + str(sample) + \"_telo_20kb.tsv\", tel_contacts, delimiter = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quantify the contacts made between chr5 telomeres with others "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "output_dir = out + \"Telo/\"\n",
    "tel_contacts_chr5 = []\n",
    "for i in [11,12]:\n",
    "    j=0\n",
    "    while j < len(coord_tel_end):\n",
    "        if i != j:\n",
    "            start1 = coord_tel_start[i]\n",
    "            end1 = coord_tel_end[i]\n",
    "            start2 = coord_tel_start[j]\n",
    "            end2 = coord_tel_end[j]\n",
    "            tel_contacts_chr5.append(sum(sum(mat_norm[start1 : end1 , start2 : end2])))\n",
    "        j += 1\n",
    "#print(tel_contacts_chr5)\n",
    "print(\"done\")\n",
    "np.savetxt(output_dir + str(sample) + \"_chr5_telo_10kb.tsv\", tel_contacts_chr5, delimiter = \"\\t\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
